File: app/src/main/java/projects/feng/gary/matrixtools/Fraction.java
package projects.feng.gary.matrixtools;

public class Fraction {
    private final int numerator;
    private final int denominator;

    public static Fraction zero = new Fraction(0, 1);
    public static Fraction one = new Fraction(1, 1);

    public Fraction(int numerator, int denominator) {
        int gcd = gcd(numerator, denominator);
        numerator /= gcd;
        denominator /= gcd;

        if (denominator < 0) {
            numerator = -numerator;
            denominator = -denominator;
        }

        this.numerator = numerator;
        this.denominator = denominator;
    }

    public Fraction(int number) {
        this.numerator = number;
        this.denominator = 1;
    }

    public Fraction(String stringRepresentation) {
        stringRepresentation = stringRepresentation.replaceAll("\\s+", "");
        String[] fractionParts = stringRepresentation.split("/");

        if (fractionParts.length == 1) {
            this.numerator = Integer.parseInt(fractionParts[0]);
            this.denominator = 1;
        } else {
            int numer = Integer.parseInt(fractionParts[0]);
            int denom = Integer.parseInt(fractionParts[1]);

            int gcd = gcd(numer, denom);

            numer /= gcd;
            denom /= gcd;

            if (denom < 0) {
                numer = -numer;
                denom = -denom;
            }

            this.numerator = numer;
            this.denominator = denom;
        }
    }

    public int getNumerator() {
        return numerator;
    }

    public int getDenominator() {
        return denominator;
    }

    public Fraction addedBy(Fraction fraction) {
        int firstNumerator = this.getNumerator() * fraction.getDenominator();
        int secondNumerator = fraction.getNumerator() * this.getDenominator();

        int newDenominator = this.getDenominator() * fraction.getDenominator();

        int newNumerator = firstNumerator + secondNumerator;

        int gcd = gcd(newNumerator, newDenominator);

        return new Fraction(newNumerator / gcd, newDenominator / gcd);
    }

    public Fraction subtractedBy(Fraction fraction) {
        return this.addedBy(fraction.negate());
    }

    public Fraction multipliedBy(Fraction fraction) {
        int newNumerator = this.numerator * fraction.numerator;
        int newDenominator = this.denominator * fraction.denominator;

        int gcd = gcd(newNumerator, newDenominator);

        return new Fraction(newNumerator / gcd, newDenominator / gcd);
    }

    public Fraction dividedBy(Fraction fraction) {
        return multipliedBy(fraction.reciprocal());
    }

    public Fraction negate() {
        return new Fraction(-this.getNumerator(), this.getDenominator());
    }

    public Fraction reciprocal() {
        return new Fraction(this.getDenominator(), this.getNumerator());
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Fraction)) {
            return false;
        }

        Fraction fraction = (Fraction) obj;
        return this.getNumerator() == fraction.getNumerator()
                && this.getDenominator() == fraction.getDenominator();
    }

    @Override
    public String toString() {
        return denominator == 1 ? String.valueOf(numerator) : this.getNumerator() + "/" + this.getDenominator();
    }

    private int gcd(int a, int b) {
        if (b == 0) {
            return a;
        }

        a = a < 0 ? -a : a;
        b = b < 0 ? -b : b;

        return gcd(b, a % b);
    }

    private int lcm(int a, int b) {
        return a * b / gcd(a, b);
    }
}

File: app/src/main/java/projects/feng/gary/matrixtools/Fraction.kt
package projects.feng.gary.matrixtools

import java.math.BigInteger


class Fraction {
    private val numerator: BigInteger
    private val denominator: BigInteger

    companion object {
        val zero = Fraction(0);
        val one = Fraction(1)
    }


    //---------------------------CONSTRUCTORS-----------------------------------------------------//

    constructor(numer: BigInteger, denom: BigInteger) {
        val gcd = if (denom < 0.toBigInteger()) -gcd(numer, denom) else gcd(numer, denom)

        numerator = numer / gcd
        denominator = denom / gcd
    }

    constructor(integer: Int) {
        numerator = integer.toBigInteger()
        denominator = 1.toBigInteger()
    }

    constructor(str: String) {
        val stringRep = str.replace("\\s".toRegex(), "")

        when {
            stringRep.equals("") -> {
                numerator = 0.toBigInteger()
                denominator = 1.toBigInteger()
            }
            stringRep.contains('/') -> {
                val fracParts = stringRep.split("/".toRegex())

                val numer = fracParts.get(0).toBigInteger()
                val denom = fracParts.get(1).toBigInteger()

                val gcd = if (denom < 0.toBigInteger()) -gcd(numer, denom) else gcd(numer, denom)

                numerator = numer / gcd
                denominator = denom / gcd
            }
            stringRep.contains('.') -> {
                val decParts = stringRep.split("\\.".toRegex())
                val numDecPoints = decParts[1].length

                // reduce fractional component before adding on the integer component
                var fracPartNum = if (numDecPoints == 0) 0.toBigInteger() else decParts[1].toBigInteger()
                var fracPartDenom = powerOfTen(numDecPoints)

                val gcd = gcd(fracPartNum, fracPartDenom)
                fracPartNum /= gcd
                fracPartDenom /= gcd

                val intPart = decParts[0].toBigInteger() * fracPartDenom

                numerator = (fracPartNum + intPart) *
                        if (decParts[0].toInt() == 0 && decParts[0].contains('-')) -1.toBigInteger() else 1.toBigInteger()
                denominator = fracPartDenom
            }
            else -> {
                numerator = stringRep.toBigInteger()
                denominator = 1.toBigInteger()
            }
        }
    }


    //---------------------------ARITHMETIC FUNCTIONS---------------------------------------------//

    operator fun plus(other: Fraction): Fraction {
        val newNumer = this.numerator * other.denominator + other.numerator * this.denominator
        val newDenom = this.denominator * other.denominator

        return Fraction(newNumer, newDenom)
    }

    operator fun minus(other: Fraction): Fraction {
        return this + -other
    }

    operator fun times(other: Fraction): Fraction {
        val newNumer = this.numerator * other.numerator
        val newDenom = this.denominator * other.denominator

        return Fraction(newNumer, newDenom)
    }

    operator fun div(other: Fraction): Fraction {
        return this * other.reciprocal()
    }

    operator fun unaryMinus(): Fraction {
        return Fraction(-this.numerator, this.denominator)
    }

    fun reciprocal(): Fraction {
        return Fraction(this.denominator, this.numerator)
    }


    //---------------------------OVERRIDES--------------------------------------------------------//

    override fun equals(other: Any?): Boolean {
        return other is Fraction
                && other.numerator == this.numerator
                && other.denominator == this.denominator
    }

    override fun toString(): String {
        return if (this.denominator == 1.toBigInteger())
            this.numerator.toString()
        else
            this.numerator.toString() + "/" + this.denominator.toString()
    }


    //---------------------------HELPERS----------------------------------------------------------//

    private fun gcd(x: BigInteger, y: BigInteger): BigInteger = if (y == 0.toBigInteger()) x.abs() else gcd(y, x.rem(y))

    private fun powerOfTen(n: Int): BigInteger {
        var result = 1

        for (i in 1..n) {
            result *= 10
        }

        return result.toBigInteger()
    }
}